tools/rosbash/rosfish

```diff
@@ -0,0 +1,357 @@
+            $EDITOR $arg
+        end
+    end
+end
+
+function roscp
+    set -l arg
+    if test  (count $argv) != "3"
+        echo -e "usage: roscp package filename target\n\nCopy a file from a package to target location."
+        return 0
+    end
+    _roscmd $argv[1] $argv[2]
+    cp $arg $argv[3]
+end
+
+function _roscomplete
+  _ros_list_locations
+  _ros_list_stacks
+end
+
+function _roscomplete_sub_dir
+    set -l wd $PWD
+    set -l roslocs (_ros_list_locations)
+    set -l arg (commandline -ct)
+    set -g rosvals
+    set -l path (_ros_decode_path $arg)
+    if test -z $rosvals[1]
+        # Not a full ROS location
+        _ros_list_locations
+    else
+        # The first section is a full ROS location, so we complete from it's subdirs
+        find -L $rosvals[2]$rosvals[3] -maxdepth 1 -mindepth 1 -type d ! -regex ".*/[.].*" -print0 | tr '\000' '\n' | sed -e "s/.*\/\(.*\)/$rosvals[1]\/\1/g"
+    end
+end
+
+function _roscomplete_rosmake
+    _roscomplete
+    set options --test-only --all --mark-installed --unmark-installed --robust --build-everything --specified-only --buildtest --buildtest1 --output --pre-clean --bootstrap --disable-logging --target --pjobs= --threads --profile --skip-blacklist --status-rate
+    for i in $options
+         echo $i
+    end
+end
+
+function _roscomplete_search_dir
+    set -l command (commandline -o)
+    if test (count $command) = 1
+      echo "1" >> ~/output.txt
+      _ros_list_packages
+    end
+    if test (count $command) = 2
+      echo "2" >> ~/output.txt
+      _ros_list_packages
+    end
+    if test (count $command) = 3
+      echo "3" >> ~/output.txt
+      find (_ros_location_find $command[1]) $arg -printf "%f\n"
+    end
+end
+
+function _roscomplete_rosrun
+    _roscomplete_search_dir '-type f -perm +111'
+end
+
+function _roscomplete_file
+    _roscomplete_search_dir '-type f ! -regex .*[.][oa]$'
+end
+
+function _roscomplete_launch
+  if test -z (commandline -ct | grep i-)
+    set options --files --args --nodes --find-node --child --local --screen --server_uri --run_id --wait --port --core --pid --dump-params
+    for i in $options
+         echo $i
+    end
+  else
+    _roscomplete_search_dir '-type f -regex \'.*\.launch$\' -o -type f -regex \'.*\.test$\''
+  end
+end
+
+complete -x -c roscd -a '(_roscomplete_sub_dir)'
+complete -c roscd -l help --description "prints command info"
+complete -c rospd -a '(_roscomplete_sub_dir)'
+complete -c rosls -a '(_roscomplete_sub_dir)'
+complete -c rosls -l help --description "prints command info"
+complete -x -c rosmake -a '(_roscomplete_rosmake)'
+complete -x -c rosclean -a 'check purge'
+complete -x -c rosrun -a '(_roscomplete_rosrun)'
+complete -x -c roslaunch -a '(_roscomplete_launch)'
+complete -x -c rosed -a '(_roscomplete_file)'
+complete -c rosed -l help --description "prints command info"
+complete -x -c roscp -a '(_roscomplete_file)'et -x TEST works
+complete -c roscp -l help --description "prints command info"
```

Comment:
Please use a consistent indentation across the whole file - either 2 or 4 spaces.